File: sdks/typescript/examples/test-secrets.ts
/**
 * Secrets Proxy Test
 *
 * Tests:
 *   1. Create a secret via the API
 *   2. Create a secret group linking the secret to an env var name
 *   3. Create a sandbox with that secret group attached
 *   4. Confirm the env var inside the sandbox shows a sealed token (osb_sealed_...), NOT the real value
 *   5. Confirm the sealed token is not the real value in any form
 *   6. Confirm the proxy substitutes the real value on outbound requests
 *      (uses httpbin.org/headers as an echo service)
 *   7. Egress allowlist blocks non-listed hosts
 *   8. Cleanup: delete secret group and secret
 *
 * Usage:
 *   npx tsx examples/test-secrets.ts
 */

import { Sandbox } from "../src/index";
import { randomBytes } from "crypto";

function green(msg: string) { console.log(`\x1b[32m\u2713 ${msg}\x1b[0m`); }
function red(msg: string) { console.log(`\x1b[31m\u2717 ${msg}\x1b[0m`); }
function bold(msg: string) { console.log(`\x1b[1m${msg}\x1b[0m`); }
function dim(msg: string) { console.log(`\x1b[2m  ${msg}\x1b[0m`); }

let passed = 0;
let failed = 0;

function check(desc: string, condition: boolean, detail?: string) {
  if (condition) {
    green(desc);
    passed++;
  } else {
    red(`${desc}${detail ? ` (${detail})` : ""}`);
    failed++;
  }
}

async function apiFetch(url: string, opts: RequestInit = {}): Promise<Response> {
  const apiKey = process.env.OPENSANDBOX_API_KEY ?? "";
  return fetch(url, {
    ...opts,
    headers: {
      "Content-Type": "application/json",
      ...(apiKey ? { "X-API-Key": apiKey } : {}),
      ...opts.headers,
    },
  });
}

async function main() {
  bold("\n\u2554\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2557");
  bold("\u2551       Secrets Proxy Test                          \u2551");
  bold("\u255a\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u255d\n");

  const apiUrl = (process.env.OPENSANDBOX_API_URL ?? "https://app.opensandbox.ai").replace(/\/+$/, "");
  const apiKey = process.env.OPENSANDBOX_API_KEY ?? "";
  if (!apiKey) {
    red("OPENSANDBOX_API_KEY is not set");
    process.exit(1);
  }

  const suffix = randomBytes(4).toString("hex");
  const secretName = `test-secret-${suffix}`;
  const groupName = `test-group-${suffix}`;
  const envVarName = "OSB_TEST_SECRET";
  const realValue = `sk-test-realvalue-${suffix}`;

  const apiBase = `${apiUrl}/api`;

  let secretId: string | null = null;
  let groupId: string | null = null;
  let sandbox: Sandbox | null = null;

  try {
    // ── Test 1: Create secret ──────────────────────────────────────────────
    bold("━━━ Test 1: Create secret ━━━\n");

    const createSecretResp = await apiFetch(`${apiBase}/secrets`, {
      method: "POST",
      body: JSON.stringify({
        name: secretName,
        description: "test secret for proxy verification",
        value: realValue,
      }),
    });
    check("Create secret returns 201", createSecretResp.status === 201,
      `got ${createSecretResp.status}: ${await createSecretResp.clone().text().then(t => t.slice(0, 100)).catch(() => "")}`);

    const secretData = await createSecretResp.json().catch(() => ({}));
    secretId = secretData.id;
    check("Secret has an ID", Boolean(secretId), JSON.stringify(secretData).slice(0, 80));
    check("Secret name matches", secretData.name === secretName, String(secretData.name));
    check("Response does NOT include the value", !("value" in secretData), Object.keys(secretData).join(", "));
    dim(`Secret: ${secretName} (id=${secretId})`);
    console.log();

    // ── Test 2: Create secret group ────────────────────────────────────────
    bold("━━━ Test 2: Create secret group ━━━\n");

    const createGroupResp = await apiFetch(`${apiBase}/secret-groups`, {
      method: "POST",
      body: JSON.stringify({
        name: groupName,
        description: "test group for proxy verification",
        allowedHosts: ["httpbin.org"],
        entries: [{ secretId, envVarName }],
      }),
    });
    check("Create secret group returns 201", createGroupResp.status === 201,
      `got ${createGroupResp.status}: ${await createGroupResp.clone().text().then(t => t.slice(0, 100)).catch(() => "")}`);

    const groupData = await createGroupResp.json().catch(() => ({}));
    groupId = groupData.id;
    check("Group has an ID", Boolean(groupId), JSON.stringify(groupData).slice(0, 80));
    check("Group name matches", groupData.name === groupName, String(groupData.name));
    dim(`Group: ${groupName} (id=${groupId})`);
    console.log();

    // ── Test 3: Create sandbox with secret group ───────────────────────────
    bold("━━━ Test 3: Create sandbox with secret group ━━━\n");

    sandbox = await Sandbox.create({
      template: "base",
      timeout: 120,
      secretGroupId: groupId!,
    });
    green(`Created sandbox: ${sandbox.sandboxId}`);
    console.log();

    // ── Test 4: Sealed token is injected (not the real value) ──────────────
    bold("━━━ Test 4: Sealed token injected ━━━\n");

    // Read from /etc/environment — canonical source
    const readEnv = await sandbox.commands.run(
      `grep '^${envVarName}=' /etc/environment | cut -d= -f2-`
    );
    const envValue = readEnv.stdout.trim().replace(/^['"]|['"]$/g, "");

    check("Env var is set in /etc/environment", Boolean(envValue), `exit=${readEnv.exitCode}`);
    check(
      "Env var is a sealed token (starts with osb_sealed_)",
      envValue.startsWith("osb_sealed_"),
      `got: ${envValue.slice(0, 60)}`,
    );
    check(
      "Real value NOT in /etc/environment",
      !readEnv.stdout.includes(realValue),
      `found real value in: ${readEnv.stdout.slice(0, 80)}`,
    );

    // Shell should also see the sealed token (not the real value)
    const shellEnv = await sandbox.commands.run(`bash -lc 'echo $${envVarName}'`);
    const shellValue = shellEnv.stdout.trim();
    check(
      "Shell sees sealed token (not real value)",
      shellValue.startsWith("osb_sealed_"),
      `got: ${shellValue.slice(0, 60)}`,
    );
    check(
      "Shell does NOT expose real value",
      !shellEnv.stdout.includes(realValue),
      "real value leaked into shell env",
    );

    const envDump = await sandbox.commands.run("env");
    check(
      "Full env dump does not contain real value",
      !envDump.stdout.includes(realValue),
      "real value found in env dump",
    );

    dim(`Sealed token: ${envValue}`);
    console.log();

    // ── Test 5: Proxy substitutes real value on outbound HTTPS ────────────
    bold("━━━ Test 5: Proxy substitutes on outbound HTTPS ━━━\n");

    // Send the sealed token as a header to httpbin.org/headers — it echoes all headers.
    // The proxy should replace the sealed token with the real value before forwarding.
    const curlCmd = `bash -lc 'curl -sf -H "x-osb-test: $${envVarName}" https://httpbin.org/headers'`;
    const curlResult = await sandbox.commands.run(curlCmd, { timeout: 30 });

    check("curl to httpbin.org succeeded", curlResult.exitCode === 0, curlResult.stderr.slice(0, 100));

    const responseBody = curlResult.stdout;
    check(
      "Response contains real value (proxy substituted the token)",
      responseBody.includes(realValue),
      `body snippet: ${responseBody.slice(0, 200)}`,
    );
    check(
      "Response does NOT contain the sealed token (proxy fully replaced it)",
      !responseBody.includes(envValue),
      `sealed token still present: ${envValue.slice(0, 30)}`,
    );

    dim(`httpbin response snippet: ${responseBody.slice(0, 150).trim()}`);
    console.log();

    // ── Test 6: Egress allowlist blocks non-listed hosts ──────────────────
    bold("━━━ Test 6: Egress allowlist blocks non-allowlisted host ━━━\n");

    const blockedCmd = `bash -lc 'curl -sf --max-time 10 https://example.com -o /dev/null -w "%{http_code}"'`;
    const blockedResult = await sandbox.commands.run(blockedCmd, { timeout: 20 });
    const blocked =
      blockedResult.exitCode !== 0 ||
      ["407", "403", "000"].includes(blockedResult.stdout.trim());
    check(
      "Non-allowlisted host is blocked by egress proxy",
      blocked,
      `exit=${blockedResult.exitCode}, stdout=${blockedResult.stdout.slice(0, 50)}`,
    );
    console.log();

    // ── Test 7: List and get operations ───────────────────────────────────
    bold("━━━ Test 7: List and get operations ━━━\n");

    const listResp = await apiFetch(`${apiBase}/secrets`);
    check("List secrets returns 200", listResp.status === 200, `got ${listResp.status}`);
    const secretsList: any[] = await listResp.json().catch(() => []);
    check("Created secret appears in list", secretsList.some(s => s.id === secretId),
      `ids: ${secretsList.slice(0, 3).map(s => s.id).join(", ")}`);
    check("List does NOT include values", secretsList.every(s => !("value" in s)), "value field leaked");

    const getGroupResp = await apiFetch(`${apiBase}/secret-groups/${groupId}`);
    check("Get secret group returns 200", getGroupResp.status === 200, `got ${getGroupResp.status}`);
    const groupDetail = await getGroupResp.json().catch(() => ({}));
    const entries: any[] = groupDetail.entries ?? [];
    check("Group has 1 entry", entries.length === 1, `got ${entries.length}`);
    check(
      "Entry has correct env var name",
      entries[0]?.envVarName === envVarName,
      JSON.stringify(entries.slice(0, 1)),
    );
    console.log();

  } catch (err: any) {
    red(`Fatal error: ${err.message || err}`);
    console.error(err);
    failed++;
  } finally {
    if (sandbox) {
      await sandbox.kill();
      green("Sandbox killed");
    }
    if (groupId) {
      await apiFetch(`${apiBase}/secret-groups/${groupId}`, { method: "DELETE" }).catch(() => {});
      green(`Secret group deleted: ${groupId}`);
    }
    if (secretId) {
      await apiFetch(`${apiBase}/secrets/${secretId}`, { method: "DELETE" }).catch(() => {});
      green(`Secret deleted: ${secretId}`);
    }
  }

  // ── Summary ──
  bold("========================================");
  bold(` Results: ${passed} passed, ${failed} failed`);
  bold("========================================\n");
  if (failed > 0) process.exit(1);
}

main().catch((err) => {
  console.error("Fatal error:", err);
  process.exit(1);
});

File: sdks/python/examples/test_secrets.py
#!/usr/bin/env python3
"""
Secrets Proxy Test

Tests:
  1. Create a secret via the API
  2. Create a secret group linking the secret to an env var name
  3. Create a sandbox with that secret group attached
  4. Confirm the env var inside the sandbox shows a sealed token (osb_sealed_...), NOT the real value
  5. Confirm the sealed token is not the real value in any form
  6. Confirm the proxy substitutes the real value on outbound requests
     (uses httpbin.org/headers as an echo service — sends sealed token as a header,
      checks the response shows the real value, proving the proxy did the swap)
  7. Cleanup: delete secret group, delete secret

Usage:
  python examples/test_secrets.py
"""

import asyncio
import os
import sys
import uuid

import httpx

from opensandbox import Sandbox

GREEN = "\033[32m"
RED = "\033[31m"
BOLD = "\033[1m"
DIM = "\033[2m"
RESET = "\033[0m"

passed = 0
failed = 0


def green(msg: str) -> None:
    print(f"{GREEN}\u2713 {msg}{RESET}")


def red(msg: str) -> None:
    print(f"{RED}\u2717 {msg}{RESET}")


def bold(msg: str) -> None:
    print(f"{BOLD}{msg}{RESET}")


def dim(msg: str) -> None:
    print(f"{DIM}  {msg}{RESET}")


def check(desc: str, condition: bool, detail: str = "") -> None:
    global passed, failed
    if condition:
        green(desc)
        passed += 1
    else:
        red(f"{desc} ({detail})" if detail else desc)
        failed += 1


async def main() -> None:
    global passed, failed

    bold("\n\u2554\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2557")
    bold("\u2551       Secrets Proxy Test                          \u2551")
    bold("\u255a\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u255d\n")

    api_url = os.environ.get("OPENSANDBOX_API_URL", "https://app.opensandbox.ai").rstrip("/")
    api_key = os.environ.get("OPENSANDBOX_API_KEY", "")
    if not api_key:
        red("OPENSANDBOX_API_KEY is not set")
        sys.exit(1)

    # Use a unique suffix so parallel test runs don't collide
    suffix = uuid.uuid4().hex[:8]
    secret_name = f"test-secret-{suffix}"
    group_name = f"test-group-{suffix}"
    env_var_name = "OSB_TEST_SECRET"
    # A realistic-looking fake API key — distinct enough to verify substitution
    real_value = f"sk-test-realvalue-{suffix}"

    api_base = f"{api_url}/api"
    headers = {"X-API-Key": api_key, "Content-Type": "application/json"}

    secret_id: str | None = None
    group_id: str | None = None
    sandbox: Sandbox | None = None

    async with httpx.AsyncClient(timeout=30) as http:
        try:
            # ── Test 1: Create secret ─────────────────────────────────────────
            bold("━━━ Test 1: Create secret ━━━\n")

            resp = await http.post(
                f"{api_base}/secrets",
                headers=headers,
                json={"name": secret_name, "description": "test secret for proxy verification", "value": real_value},
            )
            check("Create secret returns 201", resp.status_code == 201, f"got {resp.status_code}: {resp.text[:100]}")

            secret_data = resp.json()
            secret_id = secret_data.get("id")
            check("Secret has an ID", bool(secret_id), str(secret_data)[:80])
            check("Secret name matches", secret_data.get("name") == secret_name, str(secret_data.get("name")))
            check("Response does NOT include the value", "value" not in secret_data, str(list(secret_data.keys())))
            dim(f"Secret: {secret_name} (id={secret_id})")
            print()

            # ── Test 2: Create secret group ───────────────────────────────────
            bold("━━━ Test 2: Create secret group ━━━\n")

            resp = await http.post(
                f"{api_base}/secret-groups",
                headers=headers,
                json={
                    "name": group_name,
                    "description": "test group for proxy verification",
                    "allowedHosts": ["httpbin.org"],  # restrict to just httpbin for egress test
                    "entries": [{"secretId": secret_id, "envVarName": env_var_name}],
                },
            )
            check("Create secret group returns 201", resp.status_code == 201, f"got {resp.status_code}: {resp.text[:100]}")

            group_data = resp.json()
            group_id = group_data.get("id")
            check("Group has an ID", bool(group_id), str(group_data)[:80])
            check("Group name matches", group_data.get("name") == group_name, str(group_data.get("name")))
            dim(f"Group: {group_name} (id={group_id})")
            print()

            # ── Test 3: Create sandbox with secret group ──────────────────────
            bold("━━━ Test 3: Create sandbox with secret group ━━━\n")

            sandbox = await Sandbox.create(
                template="base",
                timeout=120,
                secret_group_id=group_id,
            )
            green(f"Created sandbox: {sandbox.sandbox_id}")
            print()

            # ── Test 4: Sealed token is injected (not the real value) ─────────
            bold("━━━ Test 4: Sealed token injected ━━━\n")

            # Read the env var from /etc/environment (source of truth)
            read_env = await sandbox.commands.run(
                f"grep '^{env_var_name}=' /etc/environment | cut -d= -f2-"
            )
            env_value = read_env.stdout.strip().strip("'\"")

            check("Env var is set in /etc/environment", bool(env_value), f"exit={read_env.exit_code}")
            check(
                "Env var is a sealed token (starts with osb_sealed_)",
                env_value.startswith("osb_sealed_"),
                f"got: {env_value[:60]}",
            )
            check(
                "Real value NOT in /etc/environment",
                real_value not in read_env.stdout,
                f"found real value in: {read_env.stdout[:80]}",
            )

            # Also check via shell — login shell sources /etc/environment
            shell_env = await sandbox.commands.run(f"bash -lc 'echo ${env_var_name}'")
            shell_value = shell_env.stdout.strip()
            check(
                "Shell sees sealed token (not real value)",
                shell_value.startswith("osb_sealed_"),
                f"got: {shell_value[:60]}",
            )
            check(
                "Shell does NOT expose real value",
                real_value not in shell_env.stdout,
                "real value leaked into shell env",
            )

            # env dump should never show the real value
            env_dump = await sandbox.commands.run("env")
            check(
                "Full env dump does not contain real value",
                real_value not in env_dump.stdout,
                "real value found in env dump",
            )

            dim(f"Sealed token: {env_value}")
            print()

            # ── Test 5: Proxy substitutes real value on outbound request ──────
            bold("━━━ Test 5: Proxy substitutes on outbound HTTPS ━━━\n")

            # Send the sealed token as an HTTP header to httpbin.org/headers
            # httpbin echoes all request headers in the response body.
            # If the proxy is working, the echoed value will be the REAL value, not the sealed token.
            curl_cmd = (
                f"bash -lc 'curl -sf "
                f"-H \"x-osb-test: ${env_var_name}\" "
                f"https://httpbin.org/headers'"
            )
            curl_result = await sandbox.commands.run(curl_cmd, timeout=30)

            check("curl to httpbin.org succeeded", curl_result.exit_code == 0, curl_result.stderr[:100])

            response_body = curl_result.stdout
            check(
                "Response contains real value (proxy substituted the token)",
                real_value in response_body,
                f"body snippet: {response_body[:200]}",
            )
            check(
                "Response does NOT contain the sealed token (proxy fully replaced it)",
                env_value not in response_body,
                f"sealed token still present in response: {env_value[:30]}",
            )

            dim(f"httpbin response snippet: {response_body[:150].strip()}")
            print()

            # ── Test 6: Egress allowlist blocks other hosts ───────────────────
            bold("━━━ Test 6: Egress allowlist blocks non-allowlisted host ━━━\n")

            # The group was created with allowedHosts=["httpbin.org"]
            # Attempting to reach a different host should be blocked by the proxy (407)
            blocked_cmd = (
                "bash -lc 'curl -sf --max-time 10 https://example.com -o /dev/null -w \"%{http_code}\"'"
            )
            blocked_result = await sandbox.commands.run(blocked_cmd, timeout=20)
            # curl exits non-zero on 407, or we check the status code
            blocked = (
                blocked_result.exit_code != 0
                or "407" in blocked_result.stdout
                or "403" in blocked_result.stdout
                or blocked_result.stdout.strip() in ("407", "403", "000")
            )
            check(
                "Non-allowlisted host is blocked by egress proxy",
                blocked,
                f"exit={blocked_result.exit_code}, stdout={blocked_result.stdout[:50]}, stderr={blocked_result.stderr[:50]}",
            )
            print()

            # ── Test 7: List and get group ────────────────────────────────────
            bold("━━━ Test 7: List and get operations ━━━\n")

            list_resp = await http.get(f"{api_base}/secrets", headers=headers)
            check("List secrets returns 200", list_resp.status_code == 200, f"got {list_resp.status_code}")
            secrets_list = list_resp.json()
            found = any(s.get("id") == secret_id for s in secrets_list)
            check("Created secret appears in list", found, f"ids: {[s.get('id') for s in secrets_list][:3]}")
            check("List does NOT include values", all("value" not in s for s in secrets_list), "value field leaked")

            get_group_resp = await http.get(f"{api_base}/secret-groups/{group_id}", headers=headers)
            check("Get secret group returns 200", get_group_resp.status_code == 200, f"got {get_group_resp.status_code}")
            group_detail = get_group_resp.json()
            entries = group_detail.get("entries", [])
            check("Group has 1 entry", len(entries) == 1, f"got {len(entries)}")
            check(
                "Entry has correct env var name",
                entries[0].get("envVarName") == env_var_name if entries else False,
                str(entries[:1]),
            )
            print()

        except Exception as e:
            red(f"Fatal error: {e}")
            import traceback
            traceback.print_exc()
            failed += 1

        finally:
            # Cleanup
            if sandbox:
                await sandbox.kill()
                green("Sandbox killed")
            if group_id:
                try:
                    await http.delete(f"{api_base}/secret-groups/{group_id}", headers=headers)
                    green(f"Secret group deleted: {group_id}")
                except Exception:
                    pass
            if secret_id:
                try:
                    await http.delete(f"{api_base}/secrets/{secret_id}", headers=headers)
                    green(f"Secret deleted: {secret_id}")
                except Exception:
                    pass

    # ── Summary ──
    bold("========================================")
    bold(f" Results: {passed} passed, {failed} failed")
    bold("========================================\n")
    if failed > 0:
        sys.exit(1)


if __name__ == "__main__":
    asyncio.run(main())
